### PR TITLE
[Inductor] Add decomposeK as an autotuning choice for mm

### PR DESCRIPTION
**Pull request by PaulZhang12**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150654

As a result of adding subgraph as a choice to inductor https://github.com/pytorch/pytorch/pull/149761 and enabling FP32 output from PyTorch GEMMs from FP16/BF16 inputs: https://github.com/pytorch/pytorch/pull/150812, this PR enables decompose_k as an autotuning choice for Inductor in generating the fastest matmuls with Triton. DecomposeK is currently only enabled for `torch.compile`.

Followups:
* decompose_k does not currently support epilogue fusion, which will take some work to enable
* Enable autotuning the bmm with Triton Templates as well without requiring tons of more compile time, async compilation. Anecdotal evidence shows that Triton BMM performs better usually than aten BMM
* Add for addmm
* Enable for Inference and AOTI

Below are the results of running TritonBench for Split-K shapes, comparing the aten performance versus pt2_triton, which now autotunes on decompose_k, seeing >10% speedup compared to aten on average, and for some shapes over 3x the performance of the best Triton mm previously:

Screenshot 2025-04-28 at 9 15 39 PM

TorchInductor Benchmark Dashboard:
Screenshot 2025-04-30 at 2 02 53 PM

We see speedups across all runs for training. Compile time increased as expected, with more `mm` options to tune over.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D73820115](https://our.internmc.facebook.com/intern/diff/D73820115)

### CODE REVIEW

File: test/inductor/test_max_autotune.py [hunk 1]
```diff
@@ -1128,6 +1128,34 @@ def f(x, y):
...
 
+    @parametrize("dynamic", (True, False))
+    @config.patch(
+        max_autotune=True,
+        max_autotune_gemm_backends="TRITON",
+        autotune_fallback_to_aten=False,
+    )
+    def test_max_autotune_decompose_k(self, dynamic):
+        a = torch.randn(
+            32, 32768, dtype=torch.float16, device="cuda", requires_grad=True
```
Comment (etaf) [inline test/inductor/test_max_autotune.py:1139]: Suggest to replace `"cuda"` with  `GPU_TYPE`.

File: torch/_inductor/kernel/bmm.py [hunk 2]
```diff
@@ -178,8 +189,14 @@ def may_require_contiguous(t, meta_t):
         layout,
     )
 
+    if out_dtype:
+        assert mat1.get_device().type == "cuda", "out_dtype is only supported for CUDA"
```
Comment (etaf) [inline torch/_inductor/kernel/bmm.py:193]: Hi, may I suggest you generalize this so that XPU can benefit from this?
`assert is_gpu(mat1.get_device().type)`

**Comment by PaulZhang12** [comment]
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D73820115).

File: test/inductor/test_max_autotune.py [hunk 3]
```diff
@@ -1128,6 +1128,81 @@ def f(x, y):
...
+        max_autotune_gemm_backends="TRITON",
+        autotune_fallback_to_aten=False,
+    )
+    def test_max_autotune_decompose_k(self, sizes, dynamic):
+        M, N, K = sizes
+        if K == 177147 and dynamic:
+            # decompose_k won't get selected in this case, return
+            return
+
+        a = torch.randn(M, K, dtype=torch.float16, device="cuda", requires_grad=True)
```
Comment (etaf) [inline test/inductor/test_max_autotune.py:1144]: Hi, @PaulZhang12 thanks for your reply to my last question. Since XPU have not support the dependency yet, may I suggest you mark this test case as `requires_cuda` so that the hard code `cuda` not run on XPU and crash.
 You can find the decorator by:
 `from torch.testing._internal.triton_utils import requires_cuda`

File: torch/_inductor/kernel/mm.py [hunk 4]
```diff
@@ -585,8 +592,25 @@ def has_zero_dim(size) -> bool:
 aten_bias_addmm = ExternKernelChoice(bias_addmm, None)
 
 
+def decomposeK(a, b, k_splits):
+    m = a.shape[0]
+    n = b.shape[1]
+    k = a.shape[1]
+
+    k_parts = k // k_splits
```
[the remark this answers is not in the document]
Comment (PaulZhang12) [reply torch/_inductor/kernel/mm.py:600]: Yes, we have a check `sympy.Eq(sympy.Mod(k, k_split), 0)` below before using this function

**Comment by PaulZhang12** [comment]
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D73820115).

**Review by eellison — COMMENTED** [review]
🚀 🚀 🚀 

A few things - 

- Is it worth adding a JK for enabling this if we think it's risky ? 

- I think in the first version of this PR we should avoid adding guards. This also makes it less risky.

-  Would you mind adding a test where we don't choose decompose_k ?

- I saw on the dashboard that [BertForQuestionAnswering](https://hud.pytorch.org/benchmark/huggingface/inductor_max_autotune?dashboard=torchinductor&startTime=Tue,%2022%20Apr%202025%2018:57:10%20GMT&stopTime=Tue,%2029%20Apr%202025%2018:57:10%20GMT&granularity=hour&mode=training&model=BertForQuestionAnswering&dtype=amp&deviceName=cuda%20(h100)&lBranch=gh/PaulZhang12/9/head&lCommit=9ab65e7fda2ffd9b8b03f9086db4857c28e537a5&rBranch=main&rCommit=cbcc03c2ad11fbf1080f6a1025cc3f7aee0c858d) regressed in accuracy. [AlbertForMaskedLM](https://hud.pytorch.org/benchmark/huggingface/inductor_max_autotune?dashboard=torchinductor&startTime=Tue,%2022%20Apr%202025%2018:57:10%20GMT&stopTime=Tue,%2029%20Apr%202025%2018:57:10%20GMT&granularity=hour&mode=training&model=AlbertForMaskedLM&dtype=amp&deviceName=cuda%20(h100)&lBranch=gh/PaulZhang12/9/head&lCommit=9ab65e7fda2ffd9b8b03f9086db4857c28e537a5&rBranch=main&rCommit=cbcc03c2ad11fbf1080f6a1025cc3f7aee0c858d) also had huge compilation time regression.. Could we wait to enable by default until we figure those two issues out ?

File: torch/_inductor/utils.py [hunk 5]
```diff
@@ -1522,6 +1522,39 @@ def use_cutlass_template(layout: Layout, m: int, n: int, k: int) -> bool:
     return res
 
 
+decompose_k_threshold = 16
+k_splits_limit = 5
+min_k_split, max_k_split = 16, 512
+default_k_splits = [16, 32, 64, 128, 256]
+
+
+def use_decompose_k_choice(m: int, n: int, k: int) -> bool:
```
[the remark this answers is not in the document]
Comment (PaulZhang12) [reply torch/_inductor/utils.py:1534]: Circular import, going to just add the same thing as a workaround in this file `_IntLike: TypeAlias = Union[int, sympy.Expr]
`

File: torch/_inductor/utils.py [hunk 6]
```diff
@@ -1522,6 +1522,58 @@ def use_cutlass_template(layout: Layout, m: int, n: int, k: int) -> bool:
...
+# Hand-tuned
+default_k_splits = [16, 32, 64, 128, 256]
+
+
+def use_decompose_k_choice(m: int, n: int, k: int) -> bool:
+    from torch._inductor.virtualized import V
+
+    return V.graph.sizevars.evaluate_expr(
+        sympy.Ge(k, decompose_k_threshold * m)
+    ) and V.graph.sizevars.evaluate_expr(sympy.Ge(k, decompose_k_threshold * n))
```
[the remark this answers is not in the document]
Comment (PaulZhang12) [reply torch/_inductor/utils.py:1539]: This always returns false for dynamic=True if we use `statically_known` instead of the guard. Do we want this?

File: torch/_inductor/kernel/mm.py [hunk 7]
```diff
@@ -644,6 +675,40 @@ def tuned_mm(mat1, mat2, *, layout=None):
...
+        if use_decompose_k_choice(m, n, k):
+            from torch._dispatch.python import enable_python_dispatcher
+
+            from ..decomposition import select_decomp_table
+
+            k_splits = get_k_splits(m, n, k)
+            for k_split in k_splits:
+                if not V.graph.sizevars.evaluate_expr(
+                    sympy.Eq(sympy.Mod(k, k_split), 0)
+                ):
```
[the remark this answers is not in the document]
Comment (PaulZhang12) [reply torch/_inductor/kernel/mm.py:688]: Also would always result in false

**Comment by PaulZhang12** [comment]
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D73820115).

**Comment by PaulZhang12** [comment]
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D73820115).

**Review by eellison — APPROVED** [review]

**Comment by PaulZhang12** [comment]
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D73820115).

**Comment by wdvr** [comment]
@pytorchmergebot revert -m "Failing ROCM tests: inductor/test_subgraph_choice.py::TestSubgraphChoice::test_subgraph_decompose_k [GH job link](https://github.com/pytorch/pytorch/actions/runs/14786111108/job/41515742446) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/3c54e0c21660bf6e2bf405a6aa4b9b66fd34c7b4)"

cc @PaulZhang12
